### PR TITLE
Fix coverity defects: CID 147498 147509

### PR DESCRIPTION
**Pull request by GeLiXin**
~~CID: 147498 Explicit null dereferenced~~
~~---- `mountpoints` may be NULL if we find or match nothing.~~

CID: 147509 Explicit null dereferenced
---- `l2arc_sublist_lock` is fragile as relied on caller too much.

Signed-off-by: GeLiXin ge.lixin@zte.com.cn

### CODE REVIEW

**Comment by GeLiXin** [comment]
The auto checks failed by the ztest clean up.  There is something wrong with the environment, I guess.

/usr/share/zfs/test-runner/bin/test-runner.py  -c /usr/share/zfs/runfiles/linux.run -i /usr/share/zfs/zfs-tests
Test: /usr/share/zfs/zfs-tests/tests/functional/acl/posix/setup (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/acl/posix/posix_003_pos (run as root) [00:00] [PASS]
Test: /usr/share/zfs/zfs-tests/tests/functional/acl/posix/cleanup (run as root) [01:20] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/atime/setup (run as root) [01:20] [FAIL]
Test: /usr/share/zfs/zfs-tests/tests/functional/atime/atime_001_pos (run as root) [00:00] [SKIP]

**Comment by behlendorf** [comment]
@GeLiXin every test run gets a pristine environment.  Based on fact that all the tests failed in a similar way this patch very likely introduced the issue.

**Comment by GeLiXin** [comment]
@behlendorf Thanks. I take the fix of  CID: 147498 out of this PR now,  I will find out why it make the ztest fail later.

File: module/zfs/arc.c [hunk 1]
```diff
@@ -6980,6 +6986,8 @@ l2arc_sublist_lock(int list_num)
 	case 3:
 		ml = &arc_mru->arcs_list[ARC_BUFC_DATA];
 		break;
+	default:
+		return (NULL);
```
Comment (behlendorf) [inline module/zfs/arc.c]: Also use `L2ARC_FEED_TYPES` above for the `<= 3` in the ASSERT.

File: module/zfs/arc.c [hunk 2]
```diff
@@ -7138,10 +7146,14 @@ l2arc_write_buffers(spa_t *spa, l2arc_dev_t *dev, uint64_t target_sz)
...
 	 * Copy buffers for L2ARC writing.
 	 */
-	for (try = 0; try <= 3; try++) {
+	for (try = 0; try < L2ARC_FEED_TYPES; try++) {
 		multilist_sublist_t *mls = l2arc_sublist_lock(try);
 		uint64_t passed_sz = 0;
 
+		/* Something unexpected happens. */
+		if (mls == NULL)
+			break;
```
Comment (behlendorf) [inline module/zfs/arc.c]: The comment isn't necessary here and a `continue` is more appropriate than a `break`.  This allows us to move on to the next `multilist_sublist_t`.  Although in practice none of this can ever happen.

File: lib/libzfs/libzfs_mount.c [hunk 3]
```diff
@@ -1273,6 +1273,12 @@ zpool_disable_datasets(zpool_handle_t *zhp, boolean_t force)
 	}
 
 	/*
+	 * Nothing find or match.
+	 */
+	if (mountpoints == NULL)
+		goto out;
```
Comment (behlendorf) [inline lib/libzfs/libzfs_mount.c]: Let's move this to its own patch in a different pull request.

File: module/zfs/arc.c [hunk 4]
```diff
@@ -7138,10 +7146,12 @@ l2arc_write_buffers(spa_t *spa, l2arc_dev_t *dev, uint64_t target_sz)
...
 	 * Copy buffers for L2ARC writing.
 	 */
-	for (try = 0; try <= 3; try++) {
-		multilist_sublist_t *mls = l2arc_sublist_lock(try);
+	for (try = 0; try < L2ARC_FEED_TYPES; try++) {
 		uint64_t passed_sz = 0;
 
+		multilist_sublist_t *mls = l2arc_sublist_lock(try);
+		VERIFY3P(mls, !=, NULL);
+
```
Comment (behlendorf) [inline module/zfs/arc.c]: Let's keep the change here minimal.

```
        for (try = 0; try < L2ARC_FEED_TYPES; try++) {
                multilist_sublist_t *mls = l2arc_sublist_lock(try);
                uint64_t passed_sz = 0;

                VERIFY3P(mls, !=, NULL);
```

File: lib/libzfs/libzfs_mount.c (hunk 3 above)
Comment (GeLiXin) [inline lib/libzfs/libzfs_mount.c]: This change comes back by mistake, sorry!

File: module/zfs/arc.c (hunk 4 above)
Comment (GeLiXin) [inline module/zfs/arc.c]: That's fine, I'm done~
